File: androidApp/src/main/kotlin/com/malaria/android/services/CameraService.kt
package com.malaria.android.services

import android.content.Context
import android.util.Log
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.malaria.domain.ImageInput
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.nio.ByteBuffer
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * CameraX-backed frame capture (spec §11 active screening).
 *
 * Mirror of `iosApp/Services/CameraService.swift`. Owns a `Preview` use
 * case + an `ImageAnalysis` use case bound to the host activity's
 * lifecycle. Per spec §6 ownership, the per-prediction classify task
 * runs from the composable's `rememberCoroutineScope()` — the service
 * itself only owns CameraX lifecycle and the most-recent frame.
 *
 * Capture is one-shot per tap (`captureOneFrame()`), not continuous:
 * the bound analyzer continuously updates [latestFrame] under
 * `STRATEGY_KEEP_ONLY_LATEST`, and Capture takes the buffer atomically,
 * builds a `Shared.ImageInput`, and clears the slot so the next tap
 * waits for a fresh frame rather than re-classifying the same one.
 *
 * Portrait-only per spec §11. `AndroidManifest.xml` pins the activity
 * to `android:screenOrientation="portrait"`, so the analyzer's natural
 * orientation matches.
 *
 * **YUV vs RGBA path:** the analyzer is configured with
 * [ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888] so frames arrive as a
 * single planar RGBA buffer — drop the alpha channel and we have the
 * tight-packed RGB the shared `ImageInput` expects. This avoids the
 * manual YUV→RGB matrix math that the YUV_420_888 path requires. On
 * devices that report no support for the RGBA format the builder falls
 * back to YUV; that legacy path is not exercised in v1 and a fallback
 * conversion is deliberately out of scope (would land in Phase 15
 * polish — devices that ship without RGBA output are uncommon on the
 * spec's minSdk = 36 target).
 *
 * **Emulator caveat:** the Android emulator can vend synthetic camera
 * frames, but the spec's local development setup does not have one
 * configured. End-to-end Capture is exercised on real hardware (see
 * `docs/MANUAL_TEST_PLAN.md` flow 4). Unit tests stay off the camera
 * path and only exercise the surrounding state machine
 * (`LiveOverrideStateTest`).
 */
class CameraService(
    private val context: Context,
    @Suppress("unused") private val scope: CoroutineScope,
) {

    sealed interface State {
        data object Idle : State
        data object Starting : State
        data object Running : State
        data object Stopped : State
        data class Failed(val message: String) : State
    }

    sealed class CameraError(message: String) : RuntimeException(message) {
        data object SessionNotRunning : CameraError("Camera session is not running.")
        data object CaptureTimeout : CameraError(
            "Capture timed out before a frame was available. Make sure the camera preview is running."
        )
        data class ProviderFailed(val reason: String) :
            CameraError("Camera provider failed: $reason")
    }

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state.asStateFlow()

    /**
     * Preview use case. Lazy so JVM unit tests can instantiate the
     * service without CameraX classes being initialised — the unit
     * test for `captureOneFrame()` before `start()` exercises only the
     * state-machine guard, never the camera graph. Constructed once
     * and re-bound on each `start`.
     */
    private val preview: Preview by lazy { Preview.Builder().build() }

    private val imageAnalysis: ImageAnalysis by lazy {
        ImageAnalysis.Builder()
            .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .build()
    }

    private var cameraProvider: ProcessCameraProvider? = null
    private val frameStore = LatestFrameStore()
    private val configLock = Mutex()
    private var analyzerInstalled = false

    /**
     * Wire the analyzer that forwards each frame to the lock-protected
     * store. Pulled out of the constructor so JVM unit tests can
     * instantiate the service without touching `Context`-bound
     * executors — the unit test harness's stub `android.jar` doesn't
     * vend a real `MainLooper`. Called once from [start].
     */
    private fun installAnalyzerIfNeeded() {
        if (analyzerInstalled) return
        imageAnalysis.setAnalyzer(
            ContextCompat.getMainExecutor(context),
        ) { proxy ->
            frameStore.set(proxy)
        }
        analyzerInstalled = true
    }

    // MARK: - Lifecycle

    /**
     * Bind the Preview + ImageAnalysis use cases to [lifecycleOwner].
     * Idempotent — repeated calls while already running are no-ops.
     * Throws [CameraError.ProviderFailed] if [ProcessCameraProvider]
     * initialisation fails.
     */
    suspend fun start(lifecycleOwner: LifecycleOwner) {
        val current = _state.value
        if (current is State.Running || current is State.Starting) return
        _state.value = State.Starting

        try {
            installAnalyzerIfNeeded()
            val provider = awaitCameraProvider()
            cameraProvider = provider
            withContext(Dispatchers.Main) {
                provider.unbindAll()
                provider.bindToLifecycle(
                    lifecycleOwner,
                    CameraSelector.DEFAULT_BACK_CAMERA,
                    preview,
                    imageAnalysis,
                )
            }
            _state.value = State.Running
        } catch (e: Throwable) {
            _state.value = State.Failed(e.message ?: e::class.simpleName ?: "unknown")
            throw if (e is CameraError) e else CameraError.ProviderFailed(e.message ?: "init")
        }
    }

    /**
     * Unbind every use case from the camera provider. Safe to call
     * from any thread — dispatches the actual unbind to Main. The
     * provider itself is retained for the next `start()`.
     */
    fun stop() {
        val provider = cameraProvider
        if (provider == null) {
            _state.value = State.Stopped
            return
        }
        // bindToLifecycle / unbindAll must run on Main per CameraX docs.
        ContextCompat.getMainExecutor(context).execute {
            try {
                provider.unbindAll()
            } catch (_: Throwable) {
                // CameraX raises if called pre-init; we've already
                // surfaced Failed in start(), nothing more to do.
            }
            frameStore.clear()
            _state.value = State.Stopped
        }
    }

    /**
     * Capture the next available frame from the running analyzer.
     * Polls the lock-protected store for up to ~2 seconds. The
     * STRATEGY_KEEP_ONLY_LATEST analyzer typically refreshes the slot
     * every 33 ms (30 fps), so the first poll usually succeeds.
     *
     * @throws CameraError.SessionNotRunning if `start()` hasn't run.
     * @throws CameraError.CaptureTimeout if no frame arrives in 2 s.
     */
    suspend fun captureOneFrame(): ImageInput {
        if (_state.value !is State.Running) {
            throw CameraError.SessionNotRunning
        }

        val deadlineMs = System.currentTimeMillis() + 2_000L
        while (System.currentTimeMillis() < deadlineMs) {
            val image = frameStore.takeAndBuild()
            if (image != null) return image
            // ~30 ms ≈ one frame at 30 fps.
            kotlinx.coroutines.delay(30L)
        }
        throw CameraError.CaptureTimeout
    }

    /**
     * Wire a `PreviewView`'s `SurfaceProvider` into the Preview use
     * case. Called from the [com.malaria.android.ui.home.CameraPreview]
     * composable's `AndroidView` update block on each recomposition;
     * the underlying `setSurfaceProvider` call is idempotent.
     */
    fun attachPreview(provider: Preview.SurfaceProvider) {
        preview.setSurfaceProvider(provider)
    }

    // MARK: - Provider plumbing

    /**
     * Suspend wrapper around [ProcessCameraProvider.getInstance],
     * which returns a `ListenableFuture`. The project does not depend
     * on `kotlinx-coroutines-guava` (which would let us call
     * `.await()`), so we hand-roll the listener with
     * [suspendCancellableCoroutine].
     */
    private suspend fun awaitCameraProvider(): ProcessCameraProvider = configLock.withLock {
        suspendCancellableCoroutine { cont ->
            val future = ProcessCameraProvider.getInstance(context)
            future.addListener(
                {
                    try {
                        cont.resume(future.get())
                    } catch (e: Throwable) {
                        cont.resumeWithException(
                            CameraError.ProviderFailed(e.message ?: "future"),
                        )
                    }
                },
                ContextCompat.getMainExecutor(context),
            )
            cont.invokeOnCancellation {
                // ListenableFuture cancellation is best-effort; the
                // listener above will simply see a thrown exception.
                future.cancel(false)
            }
        }
    }
}

// MARK: - Latest frame store

/**
 * Lock-protected box that owns the most-recent [ImageProxy] the
 * analyzer has produced. Single-shot semantics: [takeAndBuild]
 * consumes the stored frame, builds an [ImageInput], closes the
 * proxy, and clears the slot atomically.
 *
 * Closing the proxy is required: CameraX uses a small fixed pool of
 * buffers; leaving an `ImageProxy` un-closed back-pressures the
 * analyzer and frames stop arriving.
 */
internal class LatestFrameStore {

    private val lock = Any()
    private var latest: ImageProxy? = null

    fun set(proxy: ImageProxy) {
        val previous: ImageProxy?
        synchronized(lock) {
            previous = latest
            latest = proxy
        }
        // Close the displaced frame so the analyzer's buffer pool
        // recycles it. KEEP_ONLY_LATEST means we drop intermediates.
        previous?.close()
    }

    fun clear() {
        val previous: ImageProxy?
        synchronized(lock) {
            previous = latest
            latest = null
        }
        previous?.close()
    }

    /**
     * Atomically take the most-recent frame, build an [ImageInput]
     * from it, close the proxy, and return the input. Returns `null`
     * if no frame is queued.
     *
     * The RGBA_8888 path reads directly from `ImageProxy.planes[0]`
     * (no `proxy.image` access), so the `@ExperimentalGetImage`
     * opt-in is not required here.
     */
    fun takeAndBuild(): ImageInput? {
        val proxy: ImageProxy?
        synchronized(lock) {
            proxy = latest
            latest = null
        }
        if (proxy == null) return null
        return try {
            ImageInputBuilder.makeImageInput(proxy)
        } catch (e: Throwable) {
            Log.w("CameraService", "Failed to build ImageInput from proxy", e)
            null
        } finally {
            proxy.close()
        }
    }
}

// MARK: - Pixel buffer -> ImageInput

/**
 * Pulled out as its own type so [LiveOverrideStateTest] (and future
 * Phase 15 hardening tests) can exercise the RGBA → RGB packing
 * without spinning up CameraX. The path drops the alpha channel from
 * the RGBA_8888 buffer the analyzer is configured to produce.
 */
internal object ImageInputBuilder {

    fun makeImageInput(proxy: ImageProxy): ImageInput {
        // RGBA_8888 → single-plane buffer; rowStride may be padded
        // beyond width*4 on some devices, so honour it.
        val plane = proxy.planes[0]
        val buffer: ByteBuffer = plane.buffer
        val rowStride = plane.rowStride
        val pixelStride = plane.pixelStride
        val width = proxy.width
        val height = proxy.height

        val out = ByteArray(width * height * 3)
        val rowBytes = ByteArray(rowStride)
        var outIdx = 0
        for (row in 0 until height) {
            buffer.position(row * rowStride)
            // Read one row at a time to avoid touching the padding bytes
            // beyond `width * pixelStride`. Some buffers are direct so a
            // bulk get is cheaper than per-byte reads.
            val toRead = (width * pixelStride).coerceAtMost(rowStride)
            buffer.get(rowBytes, 0, toRead)
            var col = 0
            while (col < width) {
                val i = col * pixelStride
                out[outIdx] = rowBytes[i]         // R
                out[outIdx + 1] = rowBytes[i + 1] // G
                out[outIdx + 2] = rowBytes[i + 2] // B
                outIdx += 3
                col += 1
            }
        }

        return ImageInput(
            rgbBytes = out,
            width = width,
            height = height,
        )
    }
}

File: iosApp/Services/CameraService.swift
@preconcurrency import AVFoundation
import CoreVideo
import Foundation
import Observation
@preconcurrency import Shared
import UIKit

/// AVFoundation-backed capture service (spec §11 active screening).
///
/// Owns a single `AVCaptureSession` with a back-camera input and a video
/// data output that streams `CMSampleBuffer`s into a serial queue. The
/// view drives the lifecycle: `start()` on appear, `stop()` on disappear,
/// background, or auto-logout. Per spec §6 ownership, the per-prediction
/// classify task runs from the view's `.task`/async closure — the
/// service itself only owns the AVFoundation lifecycle and the
/// most-recent frame.
///
/// Capture is one-shot per tap (`captureOneFrame()`), not continuous:
/// the running session keeps the preview live, and Capture grabs the
/// most-recent `CVPixelBuffer` from the delegate's serial queue,
/// converts BGRA → RGB, and returns a `Shared.ImageInput`. The shared
/// `Preprocessor.resizeRGB` then downsamples to 128 at the Core ML
/// boundary; no view-side resize is required.
///
/// **Simulator caveat:** `AVCaptureSession` configuration succeeds on
/// the iOS Simulator, but Apple's simulator does not produce real
/// frames. `captureOneFrame()` will time out / never receive a
/// sample on simulator. The Capture path is exercised on real iPhone
/// hardware only. Unit tests exercise the surrounding state machine
/// and the override path without touching `AVCaptureSession`.
@Observable
@MainActor
final class CameraService {

    enum State: Sendable, Equatable {
        case idle
        case starting
        case running
        case stopped
        case failed(message: String)
    }

    enum CameraError: LocalizedError {
        case permissionDenied
        case noBackCamera
        case sessionFailed(reason: String)
        case captureTimeout

        var errorDescription: String? {
            switch self {
            case .permissionDenied:
                return "Camera access is required. Open Settings → Malaria Detector → Camera to grant access."
            case .noBackCamera:
                return "No back camera available on this device."
            case .sessionFailed(let reason):
                return "Camera session failed: \(reason)"
            case .captureTimeout:
                return "Capture timed out before a frame was available. Make sure the camera preview is running."
            }
        }
    }

    private(set) var state: State = .idle

    /// Exposed for `CameraPreviewView` to attach the preview layer.
    /// The session reference is stable across the service's lifetime;
    /// `start()` / `stop()` toggle `isRunning`, not the identity.
    nonisolated(unsafe) let session: AVCaptureSession = AVCaptureSession()

    // Background queue for blocking AVCaptureSession lifecycle calls
    // and the video-data output delegate. `startRunning()` blocks for
    // hundreds of ms — never call it on the main actor.
    private let sessionQueue = DispatchQueue(label: "camera.session", qos: .userInitiated)
    private let outputQueue = DispatchQueue(label: "camera.output", qos: .userInitiated)

    private nonisolated(unsafe) let output: AVCaptureVideoDataOutput = AVCaptureVideoDataOutput()
    private let frameStore = LatestFrameStore()
    private var sampleDelegate: SampleBufferDelegate?
    private var didConfigure = false

    init() {}

    // MARK: - Lifecycle

    func start() async throws {
        guard state != .running, state != .starting else { return }
        state = .starting

        let granted = await requestPermission()
        guard granted else {
            state = .failed(message: CameraError.permissionDenied.errorDescription ?? "denied")
            throw CameraError.permissionDenied
        }

        do {
            try await configureIfNeeded()
        } catch {
            state = .failed(message: error.localizedDescription)
            throw error
        }

        await withCheckedContinuation { (cont: CheckedContinuation<Void, Never>) in
            sessionQueue.async { [session] in
                if !session.isRunning {
                    session.startRunning()
                }
                cont.resume()
            }
        }
        state = .running
    }

    func stop() {
        sessionQueue.async { [session] in
            if session.isRunning {
                session.stopRunning()
            }
        }
        state = .stopped
    }

    /// Capture the next available frame from the running session and
    /// return it as a `Shared.ImageInput` carrying the buffer's
    /// native dimensions (BGRA → RGB on the way out).
    ///
    /// Polls `LatestFrameStore` for up to ~2 seconds. On the simulator
    /// the delegate never fires, so this throws `captureTimeout`. On
    /// real hardware the preview producing ~30fps means the first
    /// poll usually returns immediately.
    func captureOneFrame() async throws -> ImageInput {
        guard state == .running else {
            throw CameraError.sessionFailed(reason: "Camera is not running")
        }

        let deadline = Date().addingTimeInterval(2.0)
        while Date() < deadline {
            if let image = frameStore.takeAndBuild() {
                return image
            }
            try await Task.sleep(nanoseconds: 30_000_000) // ~30ms ≈ 1 frame at 30fps
        }
        throw CameraError.captureTimeout
    }

    // MARK: - Permission

    private func requestPermission() async -> Bool {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            return true
        case .notDetermined:
            return await AVCaptureDevice.requestAccess(for: .video)
        case .denied, .restricted:
            return false
        @unknown default:
            return false
        }
    }

    // MARK: - Session configuration

    /// Builds the session graph once. Subsequent `start()` calls just
    /// flip `isRunning`. Throws `noBackCamera` / `sessionFailed` if
    /// configuration cannot complete.
    private func configureIfNeeded() async throws {
        guard !didConfigure else { return }

        let delegate = SampleBufferDelegate(store: frameStore)
        sampleDelegate = delegate

        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            sessionQueue.async { [session, output, outputQueue] in
                session.beginConfiguration()
                session.sessionPreset = .photo

                guard let device = AVCaptureDevice.default(
                    .builtInWideAngleCamera,
                    for: .video,
                    position: .back
                ) else {
                    session.commitConfiguration()
                    cont.resume(throwing: CameraError.noBackCamera)
                    return
                }

                do {
                    let input = try AVCaptureDeviceInput(device: device)
                    guard session.canAddInput(input) else {
                        session.commitConfiguration()
                        cont.resume(throwing: CameraError.sessionFailed(reason: "Cannot add camera input."))
                        return
                    }
                    session.addInput(input)
                } catch {
                    session.commitConfiguration()
                    cont.resume(throwing: CameraError.sessionFailed(reason: error.localizedDescription))
                    return
                }

                output.videoSettings = [
                    kCVPixelBufferPixelFormatTypeKey as String:
                        NSNumber(value: kCVPixelFormatType_32BGRA)
                ]
                output.alwaysDiscardsLateVideoFrames = true
                output.setSampleBufferDelegate(delegate, queue: outputQueue)
                guard session.canAddOutput(output) else {
                    session.commitConfiguration()
                    cont.resume(throwing: CameraError.sessionFailed(reason: "Cannot add video data output."))
                    return
                }
                session.addOutput(output)

                if let connection = output.connection(with: .video) {
                    if connection.isVideoRotationAngleSupported(90) {
                        // 90° = portrait when the back camera is upright.
                        connection.videoRotationAngle = 90
                    }
                }

                session.commitConfiguration()
                cont.resume(returning: ())
            }
        }

        didConfigure = true
    }
}

// MARK: - Latest frame store

/// Lock-protected box that owns the most-recent CVPixelBuffer the
/// delegate has produced. Single-shot semantics: `takeAndBuild()`
/// consumes the stored buffer, converts it to a `Shared.ImageInput`,
/// and clears the slot atomically. The next Capture tap waits for a
/// fresh frame rather than re-reading an already-classified one.
///
/// CVPixelBuffer is a CoreFoundation type and is not `Sendable` in
/// strict-concurrency mode. We keep it confined behind the lock and
/// never let the raw buffer escape — only the freshly-built
/// `ImageInput` (which is plain bytes + ints) crosses the isolation
/// boundary.
final class LatestFrameStore: @unchecked Sendable {
    private let lock = NSLock()
    private var latest: CVPixelBuffer?

    func set(_ buffer: CVPixelBuffer) {
        lock.lock()
        defer { lock.unlock() }
        latest = buffer
    }

    /// Atomically take the most-recent buffer and build an
    /// `ImageInput` from it. Returns `nil` if no buffer is queued.
    func takeAndBuild() -> ImageInput? {
        lock.lock()
        let buffer = latest
        latest = nil
        lock.unlock()
        guard let buffer else { return nil }
        return try? ImageInputBuilder.makeImageInput(from: buffer)
    }
}

// MARK: - Sample buffer delegate

/// Concrete `AVCaptureVideoDataOutputSampleBufferDelegate`. The
/// delegate is invoked on the camera's outputQueue (background); it
/// just hands the pixel buffer to the lock-protected store.
final class SampleBufferDelegate: NSObject,
    AVCaptureVideoDataOutputSampleBufferDelegate, @unchecked Sendable {

    private let store: LatestFrameStore

    init(store: LatestFrameStore) {
        self.store = store
    }

    func captureOutput(
        _ output: AVCaptureOutput,
        didOutput sampleBuffer: CMSampleBuffer,
        from connection: AVCaptureConnection
    ) {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        store.set(pixelBuffer)
    }
}

// MARK: - Pixel buffer → ImageInput

/// Pulled out as its own type so `CameraServiceTests` (when they
/// land) can exercise the BGRA → RGB packing without spinning up
/// AVCaptureSession. For Phase 8 this stays internal.
enum ImageInputBuilder {

    static func makeImageInput(from pixelBuffer: CVPixelBuffer) throws -> ImageInput {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)

        guard let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer) else {
            throw CameraService.CameraError.sessionFailed(reason: "Locked pixel buffer has no base address.")
        }

        // Source: BGRA, 4 bytes per pixel.
        // Target: RGB, 3 bytes per pixel, row-packed.
        let pixelCount = width * height
        let byteArray = KotlinByteArray(size: Int32(pixelCount * 3))
        let src = baseAddress.assumingMemoryBound(to: UInt8.self)

        var outIdx: Int32 = 0
        for row in 0..<height {
            let rowStart = row * bytesPerRow
            var col = 0
            while col < width {
                let i = rowStart + col * 4
                let b = src[i]
                let g = src[i + 1]
                let r = src[i + 2]
                // Kotlin Byte is signed (Int8); reinterpret the raw bits.
                byteArray.set(index: outIdx, value: Int8(bitPattern: r))
                byteArray.set(index: outIdx + 1, value: Int8(bitPattern: g))
                byteArray.set(index: outIdx + 2, value: Int8(bitPattern: b))
                outIdx += 3
                col += 1
            }
        }

        return ImageInput(
            rgbBytes: byteArray,
            width: Int32(width),
            height: Int32(height)
        )
    }
}
